servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -150,29 +150,26 @@ public void onNext(final ServiceDiscovererEvent<ResolvedAddress> event) {
                         event);
                 final List<Host<ResolvedAddress, C>> activeAddresses =
                         activeHostsUpdater.updateAndGet(RoundRobinLoadBalancer.this, currentAddresses -> {
-                            final List<Host<ResolvedAddress, C>> refreshedAddresses =
-                                    new ArrayList<Host<ResolvedAddress, C>>(currentAddresses);
                             final ResolvedAddress addr = requireNonNull(event.address());
-                            for (int i = 0; i < refreshedAddresses.size(); i++) {
-                                Host<ResolvedAddress, C> host = refreshedAddresses.get(i);
-                                if (host.address.equals(addr)) {
-                                    if (event.isAvailable()) {
-                                        LOGGER.debug("Address {} added but it already exists.", addr);
-                                        return currentAddresses;
-                                    } else {
-                                        refreshedAddresses.remove(i);
-                                        host.markInactive();
-                                        return refreshedAddresses;
-                                    }
-                                }
-                            }
+                            final List<Host<ResolvedAddress, C>> refreshedAddresses;
                             if (event.isAvailable()) {
+                                refreshedAddresses = new ArrayList<>(currentAddresses.size() + 1);
+                                refreshedAddresses.addAll(currentAddresses);
```

Comment:
Good observation that `addAll` does a `toArray()`. However this is an implementation detail of `addAll` which isn't strictly required. This approach avoids iteration for duplicate detection and an upfront copy and additional copy/shift for removal (at the expense of iterative additions instead of bulk copy operations). I'll add benchmarks for more insight.